File: apps/api-go/internal/application/inspiration_service.go
package application

import (
	"context"
	"time"

	"github.com/minijuegosz1531/facebook_ads/apps/api-go/internal/domain/model"
	"github.com/minijuegosz1531/facebook_ads/apps/api-go/internal/domain/port"
	"github.com/minijuegosz1531/facebook_ads/apps/api-go/internal/domain/usecase"
)

// InspirationService exposes the AI pipeline to the HTTP layer.
type InspirationService struct {
	jobStore    port.JobStore
	generate    *usecase.GenerateInspirationUseCase
	attachImage *usecase.AttachImageUseCase
}

func NewInspirationService(
	adLibrary port.AdLibrary,
	imageGen port.ImageGenerator,
	copyGen port.CopyGenerator,
	storage port.Storage,
	jobStore port.JobStore,
) *InspirationService {
	return &InspirationService{
		jobStore:    jobStore,
		generate:    usecase.NewGenerateInspirationUseCase(adLibrary, imageGen, copyGen, jobStore),
		attachImage: usecase.NewAttachImageUseCase(storage, jobStore),
	}
}

// AttachImage handles a Higgsfield webhook callback for one finished image.
func (s *InspirationService) AttachImage(ctx context.Context, jobID, requestID, imageURL string) error {
	return s.attachImage.Execute(ctx, jobID, requestID, imageURL)
}

// Start persists a pending job and returns immediately, so the HTTP handler can
// reply 202 Accepted while the heavy work runs elsewhere (a goroutine in stub
// mode, an arq/Redis worker in production).
func (s *InspirationService) Start(ctx context.Context, cmd usecase.GenerateInspirationCommand) (model.InspirationJob, error) {
	job := model.InspirationJob{
		ID:        cmd.JobID,
		ClientID:  cmd.ClientID,
		Country:   cmd.Country,
		Platforms: cmd.Platforms,
		Status:    model.JobPending,
		CreatedAt: time.Now().UTC(),
	}
	if err := s.jobStore.Save(ctx, job); err != nil {
		return model.InspirationJob{}, err
	}
	return job, nil
}

// RunCommand executes the full pipeline. Callers running it in the background
// must pass a fresh context (e.g. context.Background()), NOT the HTTP request
// context — the latter is cancelled the moment the response is written.
func (s *InspirationService) RunCommand(ctx context.Context, cmd usecase.GenerateInspirationCommand) (model.InspirationJob, error) {
	return s.generate.Execute(ctx, cmd)
}

// Get returns the current state of a job for polling.
func (s *InspirationService) Get(ctx context.Context, jobID string) (model.InspirationJob, error) {
	return s.jobStore.Get(ctx, jobID)
}

File: apps/api/application/inspiration_service.py
"""Application service orchestrating the inspiration pipeline."""
from __future__ import annotations

from domain.models.inspiration import InspirationJob, JobStatus
from domain.ports.inbound.inspiration_service import IInspirationService
from domain.ports.outbound.ad_library import IAdLibrary
from domain.ports.outbound.copy_generator import ICopyGenerator
from domain.ports.outbound.image_generator import IImageGenerator
from domain.ports.outbound.job_store import IJobStore
from domain.use_cases.generate_inspiration import (
    GenerateInspirationCommand,
    GenerateInspirationUseCase,
)


class InspirationService(IInspirationService):
    def __init__(
        self,
        ad_library: IAdLibrary,
        image_generator: IImageGenerator,
        copy_generator: ICopyGenerator,
        job_store: IJobStore,
    ):
        self._job_store = job_store
        self._pending: dict[str, GenerateInspirationCommand] = {}
        self._use_case = GenerateInspirationUseCase(
            ad_library, image_generator, copy_generator, job_store
        )

    async def start(self, cmd: GenerateInspirationCommand) -> InspirationJob:
        job = InspirationJob(
            id=cmd.job_id,
            client_id=cmd.client_id,
            country=cmd.country,
            platforms=cmd.platforms,
            status=JobStatus.PENDING,
        )
        await self._job_store.save(job)
        self._pending[cmd.job_id] = cmd
        return job

    async def run(self, job_id: str) -> InspirationJob:
        cmd = self._pending.pop(job_id, None)
        if cmd is None:
            raise LookupError(f"No pending command for job {job_id}")
        return await self._use_case.execute(cmd)

    async def run_command(self, cmd: GenerateInspirationCommand) -> InspirationJob:
        """Worker entrypoint: run the pipeline directly from a serialized command."""
        return await self._use_case.execute(cmd)

    async def get(self, job_id: str) -> InspirationJob | None:
        return await self._job_store.get(job_id)
